runtime/sema/check_composite_declaration.go

```diff
@@ -1988,6 +2021,96 @@ func (checker *Checker) enumMembersAndOrigins(
 	return
 }
 
+func (checker *Checker) checkDefaultDestroyEvent(
+	eventType *CompositeType,
+	eventDeclaration ast.CompositeLikeDeclaration,
+	containerType ContainerType,
+	containerDeclaration ast.Declaration,
+) {
+
+	// an event definition always has one "constructor" function in its declaration list
+	members := eventDeclaration.DeclarationMembers()
+	functions := members.SpecialFunctions()
+	constructorFunctionParameters := functions[0].FunctionDeclaration.ParameterList.Parameters
+
+	var checkParamExpressionKind func(ast.Expression)
+	checkParamExpressionKind = func(arg ast.Expression) {
+		switch arg := arg.(type) {
+		case *ast.StringExpression,
+			*ast.BoolExpression,
+			*ast.NilExpression,
+			*ast.IntegerExpression,
+			*ast.FixedPointExpression,
+			*ast.PathExpression:
+			break
+		case *ast.IdentifierExpression:
+			// these are guaranteed to exist at time of destruction, so we allow them
+			if arg.Identifier.Identifier == SelfIdentifier || arg.Identifier.Identifier == BaseIdentifier {
+				break
+			}
+			// if it's an attachment, then it's also okay
+			if checker.typeActivations.Find(arg.Identifier.Identifier) != nil {
+				break
+			}
+			checker.report(&DefaultDestroyInvalidArgumentError{
+				Range: ast.NewRangeFromPositioned(checker.memoryGauge, arg),
+			})
+		case *ast.MemberExpression:
+			checkParamExpressionKind(arg.Expression)
+		case *ast.IndexExpression:
+			checkParamExpressionKind(arg.TargetExpression)
+			checkParamExpressionKind(arg.IndexingExpression)
+
+			// indexing expressions on arrays can fail, and must be disallowed, but
+			// indexing expressions on dicts, or composites (for attachments) will return `nil` and thus never fail
+			targetExprType := checker.Elaboration.IndexExpressionTypes(arg).IndexedType
+			switch targetExprType.(type) {
+			case *VariableSizedType, *ConstantSizedType:
+				checker.report(&DefaultDestroyInvalidArgumentError{
+					Range: ast.NewRangeFromPositioned(checker.memoryGauge, arg),
+				})
+			}
+
+		default:
+			checker.report(&DefaultDestroyInvalidArgumentError{
+				Range: ast.NewRangeFromPositioned(checker.memoryGauge, arg),
+			})
+		}
+	}
+
+	for index, param := range eventType.ConstructorParameters {
+		paramType := param.TypeAnnotation.Type
+		paramExpr := constructorFunctionParameters[index]
+		paramDefaultArgument := paramExpr.DefaultArgument
+
+		// make `self` and `base` available when checking default arguments so the fields of the composite are available
+		checker.declareSelfValue(containerType, containerDeclaration.DeclarationDocString())
```

Comment:
where (as in, in which scope) does this `self` variable get declared?

Reply:
>  I don't actually think there was any harm in declaring them like this

In the old way (i.e. defining them in the overall composite scope), could it conflict with field names of the container? (or are they defined in a separate scope? can't really remember). Maybe also add a test.